cmd/otel-allocator/allocation/allocator.go

```diff
@@ -197,41 +98,21 @@ func (allocator *Allocator) SetCollectors(collectors []string) {
 
 	allocator.m.Lock()
 	defer allocator.m.Unlock()
-	newCollectors, removedCollectors := allocator.getCollectorChanges(collectors)
-	if len(newCollectors) == 0 && len(removedCollectors) == 0 {
-		log.Info("No changes to the collectors found")
-		return
-	}
-
-	// Clear existing collectors
-	for _, k := range removedCollectors {
-		delete(allocator.collectors, k)
-		targetsPerCollector.WithLabelValues(k).Set(0)
-	}
-	// Insert the new collectors
-	for _, i := range newCollectors {
-		allocator.collectors[i] = &collector{Name: i, NumTargets: 0}
-	}
-
-	// find targets which need to be redistributed
-	var redistribute []*TargetItem
-	for _, item := range allocator.targetItems {
-		for _, s := range removedCollectors {
-			if item.Collector.Name == s {
-				redistribute = append(redistribute, item)
-			}
+	newCollectors := map[string]strategy.Collector{}
+	for _, s := range collectors {
+		newCollectors[s] = strategy.Collector{
+			Name:       s,
+			NumTargets: 0,
 		}
 	}
-	// Re-Allocate the existing targets
-	for _, item := range redistribute {
-		allocator.addTargetToTargetItems(item)
-	}
+	newState := strategy.NewState(newCollectors, allocator.state.TargetItems())
+	allocator.state = allocator.strategy.Allocate(allocator.state, newState)
 }
 
-func NewAllocator(log logr.Logger) *Allocator {
+func NewAllocator(log logr.Logger, allocatorStrategy strategy.Allocator) *Allocator {
 	return &Allocator{
-		log:         log,
-		collectors:  make(map[string]*collector),
-		targetItems: make(map[string]*TargetItem),
+		log:      log,
+		state:    strategy.NewState(make(map[string]strategy.Collector), make(map[string]strategy.TargetItem)),
```

Comment:
If we don't need/use the maps here, shouldn't the `NewState()` function be able to handle `nil` values and make the maps itself?